File: src/java/com/westlake/aird/bean/Compressor.java
/*
 * Copyright (c) 2020 CSi Biotech
 * Aird and AirdPro are licensed under Mulan PSL v2.
 * You can use this software according to the terms and conditions of the Mulan PSL v2.
 * You may obtain a copy of Mulan PSL v2 at:
 *          http://license.coscl.org.cn/MulanPSL2
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
 * See the Mulan PSL v2 for more details.
 */

package com.westlake.aird.bean;

import lombok.Data;

import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

@Data
public class Compressor {

    /**
     * 压缩策略的指向对象,目前仅支持mz对象和intensity对象两种
     */
    public static String TARGET_MZ = "mz";
    public static String TARGET_INTENSITY = "intensity";

    /**
     * 目前支持的压缩算法,包括无损的zlib,pFor算法以及有损的log10算法
     */
    public static String METHOD_ZLIB = "zlib";
    public static String METHOD_PFOR = "pFor";
    public static String METHOD_LOG10 = "log10";

    /**
     * 数组中mz和intensity的精度,1000代表精确到小数点后3位,10代表精确到小数点后1位
     */
    public static int PRECISION_MZ = 1000;
    public static int PRECISION_INT = 10;

    /**
     * 压缩对象,支持mz和intensity两种
     * Compression Target. Support for mz array and intensity array
     */
    String target;

    /**
     * 压缩方法,使用分号隔开
     * Compression Method, using comma to split the compression method with order.
     */
    List<String> methods;

    @Deprecated
    String method;

    /**
     * 数组的数值精度,10代表精确到小数点后1位,100代表精确到小数点后三位,以此类推
     */
    Integer precision;

    /**
     * ByteOrder,Aird格式的默认ByteOrder为LITTLE_ENDIAN,此项为扩展项,目前仅支持默认值LITTLE_ENDIAN
     */
    String byteOrder;

    public Integer getPrecision() {
        if (precision != null) {
            return precision;
        }

        if (target.equals(TARGET_MZ)) {
            return 1000;
        }

        if (target.equals(TARGET_INTENSITY)) {
            return 10;
        }
        return null;
    }

    public ByteOrder getByteOrder() {
        return ByteOrder.LITTLE_ENDIAN;
    }

    /**
     * 兼容性代码
     * @return
     */
    public List<String> getMethods() {
        if (methods != null) {
            return methods;
        } else if (method != null) {
            String[] methodArray = method.split(",");
            return Arrays.asList(methodArray);
        }
        return new ArrayList<>();
    }
}

File: CSharpSDK/Bean/Compressor.cs
/*
 * Copyright (c) 2020 CSi Studio
 * AirdSDK and AirdPro are licensed under Mulan PSL v2.
 * You can use this software according to the terms and conditions of the Mulan PSL v2. 
 * You may obtain a copy of Mulan PSL v2 at:
 *          http://license.coscl.org.cn/MulanPSL2 
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.  
 * See the Mulan PSL v2 for more details.
 */

using System.Collections.Generic;
using CSharpFastPFOR.Port;

namespace AirdSDK.Beans
{
    public class Compressor
    {
        public static string TARGET_MZ = "mz";
        public static string TARGET_INTENSITY = "intensity";
        public static string TARGET_MOBILITY = "mobility";
        public static string TARGET_RT = "rt";

        /**
         * Compression target, support for mz and intensity
         * 压缩对象,支持mz和intensity两种
         */
        public string target;

        /**
         * The compression method, sorted by the used order.
         * 压缩对象使用的压缩方法列表,按照顺序进行压缩
         */
        public List<string> methods;

        /**
         * Compressor precision.
         * 1000: 0.001, 3dp
         * 10000: 0.0001, 4dp
         * 100000: 0.00001, 5dp
         * 10000000: 0.0000001, 7dp
         * mz压缩精度, 有1000,10000和100000三种,代表精确到小数点后3,4,5位
         * mobility压缩精度, 目前只有10000000一种,代表精确到小数点后7位
         */
        public int precision;

        /**
         * 2^digit = layers
         * if layers = 256, then digit=8
         */
        public int digit;

        /**
         * ByteOrder,Aird格式的默认ByteOrder为LITTLE_ENDIAN,此项为扩展项,目前仅支持默认值LITTLE_ENDIAN ByteOrder
         */
        public string byteOrder;

        public Compressor(string target)
        {
            this.target = target;
        }

        public void addMethod(string method)
        {
            if (methods == null)
            {
                methods = new List<string>();
            }

            methods.Add(method);
        }

        public ByteOrder fetchByteOrder()
        {
            return ByteOrder.LITTLE_ENDIAN;
        }
    }
}